src/core/varinfo.jl

```diff
@@ -55,109 +66,158 @@ type VarInfo
+  | Logw      :   $(vi.logw)
+  | Index     :   $(vi.index)
+  | #produce  :   $(vi.num_produce)
+  \\=======================================================================
+  """
+  print(io, vi_str)
 end
 
+#################################
+# Utility functions for VarInfo #
+#################################
+
+duplicate!(vi::VarInfo) = push!(vi.vals, vi.vals[end])
+keeplast!(vi::VarInfo) = splice!(vi.vals, 1:length(vi.vals)-1)
+
+# Get all indices of variables belonging to gid or 0
+groupidcs(vi::VarInfo, gid::Int) = groupidcs(vi, gid, nothing)
+groupidcs(vi::VarInfo, gid::Int, spl::Void) = filter(i -> vi.gids[i] == gid || vi.gids[i] == 0, 1:length(vi.gids))
+groupidcs(vi::VarInfo, gid::Int, spl::Sampler) =
+  filter(i -> (vi.gids[i] == gid || vi.gids[i] == 0) && (isempty(spl.alg.space) || vi.uids[i][2] in spl.alg.space), 1:length(vi.gids))
+
 # Get all values of variables belonging to gid or 0
-groupvals(vi::VarInfo, gid::Int, spl=nothing) = map(i -> vi.vals[i], groupidcs(vi, gid, spl))
+groupvals(vi::VarInfo, gid::Int) = groupvals(vi, gid, nothing)
+groupvals(vi::VarInfo, gid::Int, spl::Union{Void, Sampler}) = map(i -> vi.vals[end][vi.ranges[i]], groupidcs(vi, gid, spl))
 
 # Get all uids of variables belonging to gid or 0
-groupuids(vi::VarInfo, gid::Int, spl=nothing) = map(i -> vi.uids[i], groupidcs(vi, gid, spl))
+groupuids(vi::VarInfo, gid::Int) = groupuids(vi, gid, nothing)
+groupuids(vi::VarInfo, gid::Int, spl::Union{Void, Sampler}) = map(i -> vi.uids[i], groupidcs(vi, gid, spl))
 
-retain(vi::VarInfo, gid::Int, n_retain, spl=nothing) = begin
-  # NOTE: the sanity check below is commented because Void
-  #       and standalone samplers uses gid = 0
-  # @assert ~(gid == 0) "[retain] wrong use of retain: gid = 0"
+# Get all uids of variables belonging to gid or 0
+getranges(vi::VarInfo, spl::Sampler) = union(map(i -> vi.ranges[i], groupidcs(vi, spl.alg.group_id, spl))...)
 
-  # Get all indices of variables belonging to gid
+retain(vi::VarInfo, gid::Int, n_retain::Int) = retain(vi, gid, n_retain, nothing)
+retain(vi::VarInfo, gid::Int, n_retain::Int, spl::Union{Void, Sampler}) = begin
   gidcs = groupidcs(vi, gid, spl)
-  l = length(gidcs)
 
-  # Remove corresponding entries
-  for i = l:-1:(n_retain + 1)
-    delete!(vi.idcs, vi.uids[gidcs[i]])
-    splice!(vi.uids, gidcs[i])
-    splice!(vi.vals, gidcs[i])
-    splice!(vi.dists, gidcs[i])
-    splice!(vi.gids, gidcs[i])
-    splice!(vi.trans, gidcs[i])
+  # Set all corresponding entries to NaN
+  l = length(gidcs)
+  for i = l:-1:(n_retain + 1),  # for each variable (in reversed order)
+      j = vi.ranges[gidcs[i]]   # for each index of variable range
+    vi[j] = NaN
   end
 
-  # Rebuild index dictionary
-  for i = 1:length(vi.uids)
-    vi.idcs[vi.uids[i]] = i
-  end
   vi
 end
 
 # Add a new entry to VarInfo
-addvar!(vi::VarInfo, vn::VarName, val, dist::Distribution, gid=0) = begin
+addvar!(vi::VarInfo, vn::VarName, val, dist::Distribution) = addvar!(vi, vn, val, dist, 0)
```

Comment:
It seems that the only place this `Var` can be used is in `initvar`, because `Var` also need `val` so it can only be used after the value is generated.